### ISSUE TITLE
[1.19.2-Fabric] AutoMap sometimes causes an internal server crash, freezing the server ticks and all entities.

### ISSUE THREAD

**Issue by Frontear** [TeamJM/journeymap#523]
#### JourneyMap Version:

1.19.2-5.9.0beta2

#### Forge Version:

Fabric 0.14.10

#### Describe the problem in repeatable steps:

1. Load into a very large world (I've used Chunky to pregen to about 5000 blocks in all directions from the origin of the world)
2. Run automap, settings don't matter from what I can see, I used defaults and it still caused this
3. Occasionally, not 100% of the time, but occasionally the server internally will crash, it will stop ticking, all entities freeze, can't interact or do anything.

#### Relevant errors from `.minecraft/journeymap/journeymap.log`:

https://mclo.gs/ndHt30r
https://streamable.com/jwnf4v

I can regenerate the jm logs if necessary, they've been overwritten by my latest playthrough, but it doesn't ever have anything meaningful.

**Comment by mysticdrew** [comment]
A couple things, You have a render error in the logs, you need to find what mod is causing that. Not sure if that is the cause to this or not. It also seems like you do not have enough memory allocated to minecraft. Look at the automap logs, it keeps having to run the java garbage collector because your memory is filling up. 

Automap is fairly memory intensive. It is also recommended to not do anything while automap is running.

**Comment by Frontear** [comment]
The GL Errors are occuring from Iris and my shaders, specifically being caused by the shaders. I usually ignore them because it doesn't affect functionality or anything visibly that I have noticed. I have upped the memory of mc to 8GB and I will still occasionally notice this happening. I don't do anything apart from letting it sit idle, and if I come back to check up on the progress I will notice its frozen.

**Comment by mysticdrew** [comment]
could you provide updated logs with more memory please?

**Comment by Frontear** [comment]
Latest logs, just skimmed through them and it looks like the JVM went OOM, which is odd since I gave it 10GB this time around. Seems to have had some kind of memory leak because memory was locked to above 95%. I can try giving it _even_ more RAM but I think id be pushing my own PC to the limits since I only have 16GB to offer. 

[journeymap.log](https://github.com/TeamJM/journeymap/files/10049564/journeymap.log)
[latest.log](https://github.com/TeamJM/journeymap/files/10049565/latest.log)

**Comment by Frontear** [comment]
Seems MC is just going OOM all the time now, which I find really really odd. Again, 10GB. There are some GL Errors, but those I've managed to track down to a mod that adds tooltips to items. For some reason it causes GL Errors, but these errors don't impact or harm functionality of the game in the slightest. 

[journeymap.log](https://github.com/TeamJM/journeymap/files/10049694/journeymap.log)
[latest.log](https://github.com/TeamJM/journeymap/files/10049695/latest.log)

**Comment by mysticdrew** [comment]
You shouldn't ever need more than 8gb allocated to MC. 

I suspect you have a mod causing memory leaks when chunks are loaded into memory. Which is why automapping causes it to happen since it loads chunks in to memory from disk while it maps it. This is not going to be an easy issue to figure out which mod is doing it. 

You will need to remove mods via binary search until you find the offending one. 
```
The binary search is a way of finding a faulty thing among a lot of other things, without having to remove the things one-by-one.
This is useful for finding a broken mod among hundreds of mods, without having to spend time testing the mods one-by-one.
The procedure is simple:
  1. Remove half of the existing things, and put them aside.
  2. Run the program / game.
  3. Does the issue still exist?
    If YES: Repeat from step 1 with the current things.
    IF NO: Swap out the current things with the ones set aside, and repeat from step 1.
  4. Repeat this process until the problematic thing/s have been found.
  ```

**Comment by Frontear** [comment]
That does seem to be the issue. Quick question, JM automap also causes chunk lighting updates, correct? How different is JM chunk loading versus if a player were to manually walk through the chunks, does it internally operate any differently?

**Comment by mysticdrew** [comment]
Automap only reads the data from disk as a ProtoChunk. It works a bit differently as you walk through them, it does not cause lighting updates. Or it shouldn't unless a mod modified the ProtoChunk loading.

**Comment by Frontear** [comment]
I did some inspection of the logs and deobfuscated the intermediary namings from Fabric, and I'm noticing these crashes occur in the ChunkLightProvider and LightStorage classes. The log entries for the second log (the ones that go OOM) point to LightStorage#updateLight, which stems from the ChunkLightProvider. Do you have any idea why the second log would have any references to lighting updates?

**Comment by Frontear** [comment]
This is the log entry, deobfuscated to the best of my ability. That being said I can confirm its definitely not these lighting updates causing an OOM, since even with Starlight enabled, which is a lighting engine rewrite mod, the OOM issues still happen. I notice MC must've gone OOM in the first log as well, and simply caused the internal server to quietly crash rather than kill the whole JVM, since I also see memory spiking higher and higher as JM goes on. Will try to cut down on the chunk loading/performance mods and see which one of them is causing it.

```
[22:00:12] [Worker-Main-4/ERROR]: Caught exception in thread Thread[Worker-Main-4,1,main]
java.lang.OutOfMemoryError: Java heap space
	at it.unimi.dsi.fastutil.longs.Long2ObjectOpenHashMap$EntryIterator.next(Long2ObjectOpenHashMap.java:803) ~[fastutil-8.5.6.jar:?]
	at it.unimi.dsi.fastutil.longs.Long2ObjectOpenHashMap$EntryIterator.next(Long2ObjectOpenHashMap.java:797) ~[fastutil-8.5.6.jar:?]
	at net.minecraft.LightStorage.updateLight(LightStorage.java:241) ~[client-intermediary.jar:?]
	at net.minecraft.ChunkLightProvider.method_15516(ChunkLightProvider.java:186) ~[client-intermediary.jar:?]
	at net.minecraft.LightingProvider.method_15516(LightingProvider.java:63) ~[client-intermediary.jar:?]
	at net.minecraft.ServerLightingProvider.runTasks(ServerLightingProvider.java:188) ~[client-intermediary.jar:?]
	at net.minecraft.ServerLightingProvider.method_17313(ServerLightingProvider.java:107) ~[client-intermediary.jar:?]
	at net.minecraft.ServerLightingProvider$$Lambda$27651/0x0000000803ecaa88.run(Unknown Source) ~[?:?]
	at net.minecraft.ChunkTaskPrioritySystem.method_17634(ChunkTaskPrioritySystem.java:62) ~[client-intermediary.jar:?]
	at net.minecraft.ChunkTaskPrioritySystem$$Lambda$27666/0x0000000803ed4f68.run(Unknown Source) ~[?:?]
	at net.minecraft.TaskExecutor.runNext(TaskExecutor.java:91) ~[client-intermediary.jar:?]
	at net.minecraft.TaskExecutor.runWhile(TaskExecutor.java:146) ~[client-intermediary.jar:?]
	at net.minecraft.TaskExecutor.run(TaskExecutor.java:102) ~[client-intermediary.jar:?]
	at java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1395) [?:?]
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373) [?:?]
	at java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182) [?:?]
	at java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655) [?:?]
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622) [?:?]
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165) [?:?]
```

**Comment by mysticdrew** [comment]
> I did some inspection of the logs and deobfuscated the intermediary namings from Fabric, and I'm noticing these crashes occur in the ChunkLightProvider and LightStorage classes. The log entries for the second log (the ones that go OOM) point to LightStorage#updateLight, which stems from the ChunkLightProvider. Do you have any idea why the second log would have any references to lighting updates?

No idea, but that OOM log you shared just shows what was happening when the JVM ran out of memory. It does not indicate what caused the memory leak. 

You may want to use a profiler and hook it to your minecraft instance. VisualVM is what I have used for memory issues.

**Comment by Frontear** [comment]
Ahh that does make a lot more sense, oversight my bad.

I will connect VisualVM and also use the spark profiler to get some insight as to whats the offending mod.

**Comment by mysticdrew** [comment]
Spark is not helpful for memory issues.

**Comment by Frontear** [comment]
Alright will use just VisualVM then. Thanks for the help so far

**Comment by Frontear** [comment]
Tried using VisualVM, I noticed there was a large increase in `byte[]` type objects, but I could not, or rather was not familiar enough with how to go about identifying them. I did obtain two heapdumps, once from before and once from minutes before an OOM crash which a pretty noticeable difference in objects in memory, so there is a mem leak. 

That being said, I tried creating a largely vanilla fabric instance with only JourneyMap, some optimization mods like Starlight, Lithium, LazyDFU, memoryLeakFix, and Chunky, and noticed that the JVM still dropped with an OOM. I also noticed back in the original mod instance, disabling all these mods and running JM still caused an OOM. Rather odd since all instances had 8GB of ram. I also took a quick ask in the MedievalMC discord (thats the modpack I am using JM on), and apparently one of the devs claims JM automap causes the OOM problem. Is that to imply JM has a mem leak somewhere?

**Comment by mysticdrew** [comment]
Can you create a brand new instance with only journeymap and fabric api. Pregen a large area and automap. I've done extensive testing and could not find any memory issues.

**Comment by mysticdrew** [comment]
>  and apparently one of the devs claims JM automap causes the OOM problem. 
 
 They have not reported it to me, they should report it to me, or comment on this issue with any details they have. It gets frustrating when people complain about issues with mods but never report it to the dev and they wonder why it never gets fixed. Thank you for taking the initiative on this. 

I am currently pregen a brand new world and will automap it when it is finished. I will report back.

**Comment by mysticdrew** [comment]
My test results so far. 
Brand new instance. 4gb memory allocated
Only Chunky (1.3.38)
Set it to 5000 blocks, started it and left the house, When I came home it had ran out of memory at 33%. Crashed with the exact same error. 144 regions still got pregened

Restarted MC with JM added back in did an automap, successfully completed memory did not go above 49% usage during automapping. 

I will try to get chunky to finish the 5000 radius and then re-automap.

**Comment by mysticdrew** [comment]
Also, feel free to join our discord it might be easier than back and forth here.

**Comment by mysticdrew** [comment]
So, I was able to get chunky to complete a 5000 radius. I think it was my fault it crashed previously. 
I automapped and it completed successfully on the first try. Though memory usage did get high at times which is expected as automap is pretty intensive, it did not come near an OOM though. 85% was about the highest usage. 

I will next add Starlight, Lithium, LazyDFU, memoryLeakFix and see what happens.

**Comment by mysticdrew** [comment]
I added the 4 mods above, automapped and it used much less memory to do it. It maxed out at 58%. 

So we seem to have something else going on here.
I used beta3 for my tests, though there should have been nothing changed that would account for less memory usage between it and beta2, but could you try running a test with beta3 please?

**Comment by Frontear** [comment]
Carried over the issue to the discord, will try to run with beta3 and report back over there

**Comment by mysticdrew** [comment]
Next release will use significantly less memory for automap! Will close when I release it

**Comment by mysticdrew** [comment]
This issue is resolved and in beta for for 1.18.2 and 1.19.3 for both fabric and forge. 
Download here:
https://www.curseforge.com/minecraft/mc-mods/journeymap/files
https://modrinth.com/mod/journeymap/versions
